board: arm: Add support for Panasonic Evaluation Boards

pideu-sj opened the pull request:
This Pull-Request adds support for Panasonic's Evaluation Boards pan1780_evb, pan1781_evb and pan1782_evb. 

  - The PAN1780 evaluation board is a development tool for the nRF52840 from Nordic Semiconductor.
  - The PAN1781 evaluation board is a development tool for the nRF52820 from Nordic Semiconductor.
  - The PAN1782 evaluation board is a development tool for the nRF52833 from Nordic Semiconductor.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj committed:
boards: Add PAN1780 evaluation board

The PAN1780 evaluation board is a development tool for the nRF52840
from Nordic Semiconductor.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj committed:
boards: Add PAN1781 evaluation board

The PAN1781 evaluation board is a development tool for the nRF52820
from Nordic Semiconductor.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj committed:
boards: Add PAN1782 evaluation board

The PAN1782 evaluation board is a development tool for the nRF52833
from Nordic Semiconductor.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

de-nordic commented:
FYI @carlescufi

pideu-mh commented:
We have noticed that the pull-request may have some problems and was not automatically merged.

We are now looking at the different reports and will go through them and try to fix them.

If you have some additional feedback for us how to proceed here, please feel free to share it with us. It's greatly appreciated.

pideu-sj commented:
Hello @de-nordic & @carlescufi,
i have a question regarding the current failures in twister-build (1) to twister-build (4), which is basically always the same devicetree error:

`devicetree error: missing data after phandle in <Property 'pwms' at '/pwmleds/pwm_evb_led_1' in '/__w/zephyr/zephyr/misc/empty_file.c'>`

We are not that much experienced with contributing to the zephyrproject and have no further ideas what could be the problem with the way we are setting the pwmleds in the devicetree file of pan1780_evb(nRF52840), pan1781_evb(nRF52820) and pan1782_evb(nRF52833). Here for example the pan1780_evb pwmleds:

`
pwmleds {
  compatible = "pwm-leds";
  pwm_evb_led1: pwm_evb_led_1 {
    pwms = <&pwm0 13>;
  };
};
`

The we way choose to add the pwmleds nodes to the devictree is exactly the way Nordic has chosen in their DK devicetree files of nrf52840dk_nrf52840, nrf52833dk_nrf52820 and nrf52833dk_nrf52833.

If you have maybe some additional information for us how to proceed, please feel free to share with us. Thank you.

gmarull reviewed (changes requested):
Thanks for your contribution, please check my comments

pideu-sj committed:
boards: Add PAN1770 evaluation board

The PAN1770 evaluation board is a development tool for the nRF52840
from Nordic Semiconductor.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,258 @@
+	pwmleds {
+		compatible = "pwm-leds";
+		pwm_evb_led1: pwm_evb_led_1 {
+			pwms = <&pwm0 13>;
```
@gmarull Thanks for the hint. I have updated the pwm in the devicetrees of our boards, and did a rebase for the pull request.

pideu-sj commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,258 @@
+	rx-pull-up;
+	rts-pin = <5>;
+	cts-pin = <7>;
+	cts-pull-up;
```
@gmarull After the new Nordic nRF Connect SDK 2.0 was released, i have updated all devicetrees of our boards, to the pinctrl handling and did a rebase for the pull request.

gmarull commented on boards/arm/pan1780_evb/board.cmake:
```diff
@@ -0,0 +1,11 @@
+
+include(${ZEPHYR_BASE}/boards/common/nrfjprog.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
```
nit: missing EOL

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/doc/index.rst:
```diff
@@ -0,0 +1,51 @@
+Overview
+********
+
+The PAN1780 Evaluation Board is a development tool for the PAN1780 module which is based on the nRF52840 chipset from Nordic Semiconductor.
+
+It is basically a clone of the official nRF52840 development kit (PCA10056) from Nordic Semiconductor. Please refer to :ref:`nrf52840dk_nrf52840` for further information.
```
please stay at 80 cols, as most rst files do

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,258 @@
+	arduino_adc: analog-connector {
+		compatible = "arduino,uno-adc";
+		#io-channel-cells = <1>;
+		io-channel-map = <0 &adc 1>,	/* A0 = P0.3 = AIN1 */
+						 <1 &adc 2>,	/* A1 = P0.4 = AIN2 */
+						 <2 &adc 4>,	/* A2 = P0.28 = AIN4 */
+						 <3 &adc 5>,	/* A3 = P0.29 = AIN5 */
+						 <4 &adc 6>,	/* A4 = P0.30 = AIN6 */
+						 <5 &adc 7>;	/* A5 = P0.31 = AIN7 */
```
nit: please fix alignment here

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,258 @@
+		#gpio-cells = <2>;
+		gpio-map-mask = <0xffffffff 0xffffffc0>;
+		gpio-map-pass-thru = <0 0x3f>;
+		gpio-map = <0 0 &gpio0 3 0>, /* A0 */
+					<1 0 &gpio0 4 0>, /* A1 */
+					<2 0 &gpio0 28 0>, /* A2 */
+					<3 0 &gpio0 29 0>, /* A3 */
```
nit: please fix alignment here

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,258 @@
+
+&i2c1 {
+	compatible = "nordic,nrf-twi";
+	/* Cannot be used together with spi1, so disabled by default. */
+	status = "disabled";
```
nit: I think SoC dtsi files already flag this one as disabled, so a comment should be sufficient.

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,258 @@
+
+&spi0 {
+	compatible = "nordic,nrf-spi";
+	/* Cannot be used together with i2c0, so disabled by default. */
+	status = "disabled";
```
ditto

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/pan1780_evb_defconfig:
```diff
@@ -0,0 +1,44 @@
+# additional board options
+CONFIG_GPIO_AS_PINRESET=y
+
+# Please activate NFC following NFC config on pan1780_evb
+#CONFIG_NFC_T2T_NRFXLIB=y
+#CONFIG_NFC_NDEF=y
+#CONFIG_NFC_NDEF_MSG=y
+#CONFIG_NFC_NDEF_RECORD=y
+#CONFIG_NFC_NDEF_URI_MSG=y
+#CONFIG_NFC_NDEF_URI_REC=y
+
+# bluetooth advertising name
+CONFIG_BT_DEVICE_NAME="pan1780_evb"
```
boards only enable minimal peripherals, these options seem to belong to the application domain

pideu-sj replied:
Fixed

gmarull commented on boards/arm/pan1780_evb/pan1780_evb_defconfig:
```diff
@@ -0,0 +1,44 @@
+CONFIG_BT_DEVICE_NAME="pan1780_evb"
+
+# using pinctrl
+CONFIG_PINCTRL=y
```
missing EOL

pideu-sj replied:
Fixed

gmarull reviewed (changes requested):
Thanks for the update, overall LGTM just some minor comments (apply to all boards)

pideu-sj commented:
Hi @gmarull thanks for the review, i will step through your comments and fix them.

But i have already a question regarding the twister run. After updating pinctrl handling and pwmleds, almost all twister runs are successful. Nevertheless, there is a new "board not supported" error in the "test_adc.c" file, when testing the "zephyr\tests\drivers\adc\adc_api\". 

When i'm looking into "test_adc.c", there is of course no CONFIG_BOARD_PAN1770_EVB. Do i have to add my boards on my own to the test and commit afterwards?

gmarull commented:
> Hi @gmarull thanks for the review, i will step through your comments and fix them.
> 
> But i have already a question regarding the twister run. After updating pinctrl handling and pwmleds, almost all twister runs are successful. Nevertheless, there is a new "board not supported" error in the "test_adc.c" file, when testing the "zephyr\tests\drivers\adc\adc_api".
> 
> When i'm looking into "test_adc.c", there is of course no CONFIG_BOARD_PAN1770_EVB. Do i have to add my boards on my own to the test and commit afterwards?

The test_adc.c seems to have still hardcoded information on a board basis, so yes, you'll need to manually add your board there. This tests likely needs a refactor now that ADC props can be configured in Devicetree, as its current structure doesn't scale. We recently refactored https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc in this direction, you can add overlays for your board there as well.

pideu-sj committed:
boards: Add board config defines to adc test

The test_adc.c still have hardcoded information on a board basis,
that need to be added manually.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented:
> 

@gmarull Thanks for the hint again. I've fixed the issues and did a rebase on the pull request.

gmarull commented on boards/arm/pan1770_evb/pan1770_evb.dts:
```diff
@@ -0,0 +1,257 @@
+		#gpio-cells = <2>;
+		gpio-map-mask = <0xffffffff 0xffffffc0>;
+		gpio-map-pass-thru = <0 0x3f>;
+		gpio-map = <0 0 &gpio0 3 0>,   /* A0 */
+				   <1 0 &gpio0 4 0>,   /* A1 */
+				   <2 0 &gpio0 28 0>,  /* A2 */
+				   <3 0 &gpio0 29 0>,  /* A3 */
+				   <4 0 &gpio0 30 0>,  /* A4 */
+				   <5 0 &gpio0 31 0>,  /* A5 */
+				   <6 0 &gpio1 1 0>,   /* D0 */
+				   <7 0 &gpio1 2 0>,   /* D1 */
+				   <8 0 &gpio1 3 0>,   /* D2 */
+				   <9 0 &gpio1 4 0>,   /* D3 */
+				   <10 0 &gpio1 5 0>,  /* D4 */
+				   <11 0 &gpio1 6 0>,  /* D5 */
+				   <12 0 &gpio1 7 0>,  /* D6 */
+				   <13 0 &gpio1 8 0>,  /* D7 */
+				   <14 0 &gpio1 10 0>, /* D8 */
+				   <15 0 &gpio1 11 0>, /* D9 */
+				   <16 0 &gpio0 12 0>, /* D10 */
+				   <17 0 &gpio0 13 0>, /* D11 */
+				   <18 0 &gpio0 14 0>, /* D12 */
+				   <19 0 &gpio0 15 0>, /* D13 */
+				   <20 0 &gpio0 26 0>, /* D14 */
+				   <21 0 &gpio0 27 0>; /* D15 */
+	};
+
+	arduino_adc: analog-connector {
+		compatible = "arduino,uno-adc";
+		#io-channel-cells = <1>;
+		io-channel-map = <0 &adc 1>,	/* A0 = P0.3 = AIN1 */
+						 <1 &adc 2>,	/* A1 = P0.4 = AIN2 */
+						 <2 &adc 4>,	/* A2 = P0.28 = AIN4 */
+						 <3 &adc 5>,	/* A3 = P0.29 = AIN5 */
+						 <4 &adc 6>,	/* A4 = P0.30 = AIN6 */
+						 <5 &adc 7>;	/* A5 = P0.31 = AIN7 */
```
nit: alignment is still wrong here, is your editor configured correctly? tab width should be 8

pideu-sj replied:
Oh that's strange, i thought i got the correct tab size now and it looked good in my editor. I will check this again. Thanks!

pideu-sj replied:
Tab width is now definitely set to 8. So it's looking good in all of my editors. And I took the clean nRF52840dk_nrf52840.dts from zephyr repo as an example, which using the same tabs and spaces and looks also good in editor.

pideu-sj commented:
Hi @gmarull 

I noticed, that after last rebase of the pull request, the twister runs 2 & 3 are still failed. After looking for details, i can see, that some shield tests are failing, because our "arduino_header: connector" in the dts files for pan1781_evb and pan1782_evb is not able, to deliver a full gpio-map. This is simply, because the nRF52833 and nRF52820 SOC's have not enough pins to fill the gpio-map.

What would you suggest in this case? Is it possible, to exclude these shield tests from the twister run and live with the fact, that the shields will not be compatible with our evaluation board? If yes, how can this be done?

gmarull commented on boards/arm/pan1781_evb/pan1781_evb_defconfig:
```diff
@@ -0,0 +1,36 @@
+# additional board options
+CONFIG_GPIO_AS_PINRESET=y
+
+# bluetooth advertising name
+CONFIG_BT_DEVICE_NAME="pan1781_evb"
```
this should not be at board level, applications need to choose their name.

pideu-sj replied:
Yes i have already seen the warning, when building an application, with this configuration. The reason, why i have added  this configuration to the *_defconfig file is that all of our boards should initially run with the same software. When i add the advertising name to the application, all of our boards would advertise with the same name. So our solution was to add this board specific naming to the files that define the boards, so the *_defconfig configuration. Hope that way of configuration is okay for you.

gmarull replied:
The board defconfigs need to be kept minimal per porting guidelines, this option is clearly not minimal (depends on BT). You can add it to your application `prj.conf`, see for example https://github.com/teslabs/zds-2022-drivers-app/blob/main/app/prj.conf#L19

pideu-sj replied:
Yes, you are right. I can also handle this on application level, by asking for the CONFIG_BOARD_PAN1780_EVB. I will remove the CONFIG_BT_DEVICE_NAME from *_defconfig files.

gmarull commented on boards/arm/pan1780_evb/pan1780_evb_defconfig:
```diff
@@ -0,0 +1,36 @@
+# additional board options
+CONFIG_GPIO_AS_PINRESET=y
+
+# bluetooth advertising name
+CONFIG_BT_DEVICE_NAME="pan1780_evb"
```
ditto

gmarull commented on boards/arm/pan1781_evb/pan1781_evb.dts:
```diff
@@ -0,0 +1,150 @@
+		};
+	};
+
+	arduino_header: connector {
+		compatible = "arduino-header-r3";
+		#gpio-cells = <2>;
+		gpio-map-mask = <0xffffffff 0xffffffc0>;
+		gpio-map-pass-thru = <0 0x3f>;
+		gpio-map = <0 0 &gpio0 2 0>,   /* A0 */
+				   <1 0 &gpio0 3 0>,   /* A1 */
+				   <2 0 &gpio0 4 0>,   /* A2 */
+				   <3 0 &gpio0 5 0>,   /* A3 */
+				   <6 0 &gpio0 6 0>,   /* D0 */
+				   <7 0 &gpio0 8 0>,   /* D1 */
+				   <8 0 &gpio0 14 0>,  /* D2 */
+				   <9 0 &gpio0 15 0>,  /* D3 */
+				   <10 0 &gpio0 4 0>,  /* D4 */
+				   <11 0 &gpio0 5 0>,  /* D5 */
+				   <16 0 &gpio0 3 0>,  /* D10 */
+				   <17 0 &gpio0 20 0>, /* D11 */
+				   <18 0 &gpio0 29 0>, /* D12 */
+				   <19 0 &gpio0 28 0>, /* D13 */
+				   <20 0 &gpio0 7 0>,  /* D14 */
+				   <21 0 &gpio0 30 0>; /* D15 */
+	};
```
if it has some entries missing, maybe it should not be considered an `arduino-header-r3` then? cc: @anangl

pideu-sj replied:
Hmm okay. Yeah maybe you are right. When the `arduino-header-r3 ` is defined like it is in "arduino-header-r3.yaml" everyone (including the twister tests) assumes, that the chip is really connected to all of the A0-A5 and the D0-D15 pins.
On the other hand, it's really unpleasant that we have to remove the whole header now, because just some of them are missing. But if iam got you right, there is no other option to succeed the twister runs and finally complete the upload of our boards? So if there is no arduino header present in the device tree, the shield tests will not be active for the twister runs on our board?

gmarull replied:
I guess one solution is to create a compatible specific to your connector variant, and update the `samples/subsys/display/lvgl/sample.yaml` so that it depends on connector r3 compatible for the `sample.display.lvgl.adafruit_2_8_tft_touch_v2` test.

that is, add:

```
filter: dt_compat_enabled("arduino-connector-r3")
```

gmarull replied:
FYI, I have started fixing some tests: https://github.com/zephyrproject-rtos/zephyr/pull/46835. Note that you should create a new compatible for your connector, I see no other way to reliably exclude tests that expect a complete arduino R3 connector. Adding this board to the exclude list because it misses one pin is not maintainable, as it will continue impacting new additions to the tree. For now, the compatible won't do anything until tests are improved, so your exclude list needs to be kept.

pideu-sj replied:
Hi @gmarull 
Now I have created the compatible `panasonic,reduced-arduino-header` and changed the compatible for the pan1782_evb and pan1782_evb in the dts files. These used the arduino-header-r3 before. Hope that fits for you now?

pideu-sj committed:
boards: Exclude boards from shield samples

Some of the Panasonic's evaluation boards have limited amount of pins.
So not all arduino pins can be occupied and not all shields are
compatible.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented on boards/arm/pan1780_evb/pan1780_evb.dts:
```diff
@@ -0,0 +1,256 @@
+		gpio-map = <0 0 &gpio0 3 0>,   /* A0 */
+				   <1 0 &gpio0 4 0>,   /* A1 */
+				   <2 0 &gpio0 28 0>,  /* A2 */
+				   <3 0 &gpio0 29 0>,  /* A3 */
```
Tab width is now definitely set to 8. So it's looking good in all of my editors. And I took the clean nRF52840dk_nrf52840.dts from zephyr repo as an example, which using the same tabs and spaces and looks also good in editor.

gmarull commented on boards/arm/pan1781_evb/pan1781_evb.dts:
```diff
@@ -0,0 +1,150 @@
+		};
+	};
+
+	pwmleds {
+		compatible = "pwm-leds";
+		pwm_evb_led1: pwm_evb_led_1 {
+			pwms = <&sw_pwm 0 PWM_MSEC(20) PWM_POLARITY_INVERTED>;
+		};
+	};
```
where is `sw_pwm` enabled?

pideu-sj replied:
Yes you are right, i can see that `sw_pwm` node in `nrf52820.dtsi` is disabled by default, and it will not be enabled on another file. But our template for generation of this devicetree was the file `nRF52833dk_nrf52820.dts` from Nordic. Here it was done exactly the same way, we have done. What would you suggest? Do i have to add the `status = "okay";` to the node, before i get an go from you?

gmarull replied:
yes, and you also need to setup to which GPIO signal needs to be generated, see e.g. `boards/arm/nrf51dk_nrf51422/nrf51dk_nrf51422.dt`

pideu-sj replied:
Ah okay. Thanks a lot for this hint to `boards/arm/nrf51dk_nrf51422/nrf51dk_nrf51422.dts`. I have updated the device tree like you've suggested.

pideu-sj commented on samples/subsys/fs/fat_fs/sample.yaml:
```diff
@@ -14,7 +14,7 @@ tests:
     platform_allow: nrf52840_blip
   sample.filesystem.fat_fs.adafruit_2_8_tft_touch_v2:
     depends_on: arduino_spi arduino_gpio arduino_i2c
-    platform_exclude: stm32f769i_disco
```
Thanks a lot.

pideu-sj committed:
boards: Add Panasonic header gpio binding

Panasonic's reduced Arduino header is based on the Arduino UNO layout.
The main difference is that some pins are not available due to the
lack of pins on the Panasonic module.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented on boards/arm/pan1770_evb/pan1770_evb-pinctrl.dtsi:
```diff
@@ -0,0 +1,184 @@
+
+	spi3_default: spi3_default {
+		group1 {
+			psels = <NRF_PSEL(SPIM_SCK, 0, 15)>,
+				<NRF_PSEL(SPIM_MISO, 0, 14)>,
+				<NRF_PSEL(SPIM_MOSI, 0, 13)>;
```
Hi @anangl 
there is an Header P20 available on the evaluation board, that can disconnect the `evb_led1` - `evb_led4`, from the pins `P0.13` - `P0.16`. Please see page 6 in the schematics: ([https://pideu.panasonic.de/development-hub/pan1780/downloads/PAN1780%20EVB%20Schematics.pdf](url)). So after removing the jumpers, only the `arduino_spi` is left connected. Is this a problem, to already assign both functions to the pins in the devicetree?
And you are right, `P0.13` is also assigned as `pwm0`, but that was what we wanted. So the `evb_led1` can also be used for the basic zephyr/Nordic examples as an pwm led. And I can see that Nordic does the same for `P0.13` on their `nRF52840dk_nrf52840` development kit.

anangl replied:
Okay, I think that `spi3` should be disabled by default then. And it would be nice to add some information that it cannot be used together with LEDs 1-3, SW2 button, and `pwm0` (a comment in the `spi3` node should be enough, I think). Otherwise, users can be confused.
It's not a problem if the same pins are assigned to more than one node in devicetree (as you do also for I2C/SPI and SPI/QSPI, what is perfectly fine), a problem would be if a user would try to use the corresponding devices simultaneously.

pideu-sj replied:
Yes thanks! Good idea. Totally agree! A comment in `spi3` will make that much clearer for the user. I will add this comment and update the pull request commits for pan1780_evb and pan1770_evb.

pideu-sj replied:
Requested change is now pushed.

anangl commented on boards/arm/pan1781_evb/pan1781_evb.dts:
```diff
@@ -0,0 +1,157 @@
+
+&sw_pwm {
+	status ="okay";
+	channel-gpios = <&gpio0 14 0>;
```
```suggestion
	channel-gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
```

pideu-sj replied:
Updated to your suggestion. Thanks.

gmarull commented:
@pideu-sj for bisectability, last commit should go first

pideu-sj commented:
> @pideu-sj for bisectability, last commit should go first



> @pideu-sj for bisectability, last commit should go first

Changed commit order.
